### ISSUE TITLE
Additional Breaking Changes for Dart 3.0

### ISSUE THREAD

**Issue by TimWhiting** [dart-lang/sdk#49892]
I believe Dart 3.0 has the opportunity to make more breaking changes than just only disallowing non-null-safe code.

For example, I have seen in several issues that the dart language team has said, 'if we had null-safety to start with, we would have designed this api differently'.

Feel free to edit or add additional breaking changes to consider / discuss, but here are a few I can think of.

- [ ] collections with StateError `list.first`, `list.last`, `list.firstWhere` -> change return type to nullable. Migrate existing usages to `List.first.. ?? throw ..`. This has the benefit of making it much more explicit to the user where their code can throw. I know I've been bitten by these methods in the past.
- [ ] MapEntry -> `typedef MapEntry<K,V> = ({K key, V value})` using the Records feature which seems to be progressing quickly. This only breaks implementors/extenders of MapEntry I believe unless someone is relying on identity of the MapEntry object. Searching Github for "implements MapEntry" I see 95 usages in Dart code. (https://github.com/search?l=Dart&p=7&q=%22implements+MapEntry%22&type=Code) Most are copies of the `LruCache`'s `_Link` from `package:build` or SplayTreeMapNode it looks like, but there are a few which basically just expose another name for the key or value, these can be migrated to extension methods or 'Views'. I'm not sure how to handle the migration of `_Link` or `SplayTreeMapNode` admittedly.
- [ ] Add `.indexed` getter on List (could just be an extension method?) that returns a Record `(int index, T value)`

**Comment by lrhn** [comment]
Platform library changes are very difficult to pull off. Worst case, we need to migrate the world *at the same time* that we make the change. APIs do not have language versioning, so we can't change behavior based on the language version of the code calling the method.

Convering `MapEntry` to a record (or maybe a [#struct](https://github.com/dart-lang/language/issues/2360)) is something we have considered. I'd probably prefer the struct (or view on record) so that it has its own type.

The `.indexed` getter on `List` would definitely have to be an extension, and then the only thing holding anyone back from doing it today is the lack of a record type to use for the value. (But `.asMap().entries` is an `Iterable<MapEntry<int, T>>`, so maybe that's just what you want).

**Comment by bernaferrari** [comment]
The MapEntry idea is awesome. Whatever you decide, I'm in.

**Comment by lrhn** [comment]
The problem with making `MapEntry` a typedef of a record type is that it won't have the constructor that people are currently using. You can't do `MapEntry(key, value)` any more.
A `view` on a record could probably have a constructor, and the `struct` design would be the closest thing to the current class, while still not requiring a persistent identity. So, those are more likely to succeed, since the migration will be much smaller.

**Comment by bernaferrari** [comment]
Yeah, makes sense. I'm not familiar with the view, if it works, perfect. If it doesn't, dart fix wouldn't be hard.

**Comment by bernaferrari** [comment]
As much as I want union types (personal favorite request), I don't think it will arrive before views/other things.

**Comment by ykmnkmi** [comment]
Variable number of arguments `varargs`.
From `package:path` changelog:
```
1.8.3: Support up to 16 arguments in join function and up to 15 arguments in absolute function.
```

**Comment by bernaferrari** [comment]
If I could ask for anything crazy in Dart 3, I would go with `print("value is", a)`. I alwaaays struggle with Dart syntax, and it is possible to implement that without varargs. Since print is not supposed to run in release mode, the minimal performance impact is negligible.

https://github.com/dart-lang/sdk/issues/49084

**Comment by dupuchba** [comment]
> 

@TimWhiting 
well, the ClojureDart compiler does use `MapEntry` and while migrating to Dart 3 we're going to loose this interop btw Dart ME and clojure ME.

**Comment by TimWhiting** [comment]
These were just proposed changes. I'm not part of the dart team, and the map entry change is unlikely to happen at least for dart 3.

**Comment by lrhn** [comment]
The `MapEntry` class was made `final` in the Dart 3 platform libraries, which means it cannot be subclassed by other 3.0 libraries.
So that has happened.

It won't break existing 2.19 code, but it may require some rewriting when upgrading to language version 3.0.
